File: round-table/sdk/python/src/roundtable/messages.py
# Copyright (c) 2025 AInTandem
# SPDX-License-Identifier: MIT

"""
Message client for Round Table SDK.

Provides methods for sending and managing messages between agents.
"""

from __future__ import annotations

import logging
from typing import Any

from roundtable.models import (
    AgentMessage,
    MessageListResponse,
)

logger = logging.getLogger(__name__)


class MessageClient:
    """
    Client for message operations.

    Provides methods for sending messages, getting messages,
    and broadcasting to workspaces.
    """

    def __init__(self, http_client: Any):
        """
        Initialize the message client.

        Args:
            http_client: HTTP client instance
        """
        self._client = http_client

    async def send(
        self,
        from_sandbox_id: str,
        to_sandbox_id: str,
        content: dict[str, Any],
        message_type: str = "request",
    ) -> AgentMessage:
        """
        Send a message from one sandbox to another.

        Args:
            from_sandbox_id: Sender sandbox ID
            to_sandbox_id: Recipient sandbox ID
            content: Message content
            message_type: Type of message (request, response, notification, etc.)

        Returns:
            Sent AgentMessage

        Raises:
            NotFoundError: If sandbox doesn't exist
            RoundTableError: If the request fails
        """
        request_data = {
            "to_sandbox_id": to_sandbox_id,
            "content": content,
            "message_type": message_type,
        }

        response = await self._request(
            "POST",
            f"/messages/sandboxes/{from_sandbox_id}/messages",
            json=request_data,
        )
        data = response["data"]

        return AgentMessage(**data)

    async def get_messages(
        self,
        sandbox_id: str,
        offset: int = 0,
        limit: int = 100,
    ) -> MessageListResponse:
        """
        Get messages for a sandbox.

        Args:
            sandbox_id: Sandbox ID
            offset: Pagination offset
            limit: Maximum number of messages to return

        Returns:
            MessageListResponse with list of messages

        Raises:
            NotFoundError: If sandbox doesn't exist
            RoundTableError: If the request fails
        """
        response = await self._request(
            "GET",
            f"/messages/sandboxes/{sandbox_id}/messages",
            params={"offset": offset, "limit": limit},
        )
        data = response["data"]

        messages = [AgentMessage(**msg) for msg in data["messages"]]

        return MessageListResponse(
            messages=messages,
            count=data["count"],
            offset=data.get("offset", offset),
            limit=data.get("limit", limit),
        )

    async def get_message(self, message_id: str) -> AgentMessage:
        """
        Get message details by ID.

        Args:
            message_id: Message ID

        Returns:
            AgentMessage details

        Raises:
            NotFoundError: If message doesn't exist
            RoundTableError: If the request fails
        """
        response = await self._request(
            "GET",
            f"/messages/messages/{message_id}",
        )
        data = response["data"]

        return AgentMessage(**data)

    async def broadcast(
        self,
        workspace_id: str,
        content: dict[str, Any],
        message_type: str = "notification",
    ) -> dict[str, Any]:
        """
        Broadcast a message to all sandboxes in a workspace.

        Args:
            workspace_id: Workspace ID
            content: Message content
            message_type: Type of message

        Returns:
            Broadcast response data

        Raises:
            NotFoundError: If workspace doesn't exist
            RoundTableError: If the request fails
        """
        request_data = {
            "content": content,
            "message_type": message_type,
        }

        response = await self._request(
            "POST",
            f"/messages/workspaces/{workspace_id}/broadcast",
            json=request_data,
        )

        return response["data"]

    async def _request(
        self,
        method: str,
        path: str,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """
        Make an HTTP request.

        Args:
            method: HTTP method
            path: Request path
            **kwargs: Additional arguments

        Returns:
            Response data
        """
        url = path if path.startswith("http") else path

        response = await self._client.request(method, url, **kwargs)
        data = response.json()

        if not data.get("success") and response.status_code < 400:
            raise Exception(f"Request failed: {data.get('message')}")

        return data

File: round-table/sdk/typescript/src/messages.ts
// Copyright (c) 2025 AInTandem
// SPDX-License-Identifier: MIT

import { AxiosInstance } from 'axios';
import {
  AgentMessage,
  MessageCreateRequest,
  MessageListResponse,
  BroadcastRequest,
} from './types';

/**
 * Client for message operations
 *
 * Provides methods for sending messages between agents and broadcasting.
 */
export class MessageClient {
  constructor(private readonly httpClient: AxiosInstance) {}

  /**
   * Send a message from one sandbox to another
   *
   * @param fromSandboxId - Sender sandbox ID
   * @param request - Message creation request
   * @returns Sent message
   */
  async send(fromSandboxId: string, request: MessageCreateRequest): Promise<AgentMessage> {
    const response = await this.httpClient.post(
      `/sandboxes/${fromSandboxId}/messages`,
      request
    );
    return response.data.data;
  }

  /**
   * Get messages for a sandbox
   *
   * @param sandboxId - Sandbox ID
   * @param params - Pagination parameters
   * @returns Message list response
   */
  async getMessages(
    sandboxId: string,
    params?: { offset?: number; limit?: number }
  ): Promise<MessageListResponse> {
    const response = await this.httpClient.get(`/sandboxes/${sandboxId}/messages`, {
      params: {
        offset: params?.offset ?? 0,
        limit: params?.limit ?? 100,
      },
    });

    return response.data.data;
  }

  /**
   * Get message by ID
   *
   * @param messageId - Message ID
   * @returns Message details
   */
  async get(messageId: string): Promise<AgentMessage> {
    const response = await this.httpClient.get(`/messages/${messageId}`);
    return response.data.data;
  }

  /**
   * Broadcast a message to all sandboxes in a workspace
   *
   * @param workspaceId - Workspace ID
   * @param request - Broadcast request
   * @returns Broadcast response data
   */
  async broadcast(workspaceId: string, request: BroadcastRequest): Promise<{
    broadcast_to: number;
    workspace_id: string;
  }> {
    const response = await this.httpClient.post(
      `/workspaces/${workspaceId}/broadcast`,
      request
    );
    return response.data.data;
  }
}
